router/streams2.ts

```diff
@@ -0,0 +1,310 @@
+   * but the stream is fully consumed and closed. We don't need
+   * to signal that break was called, only that the stream is closed.
+   */
+  private didBreakDisposeValues = false;
+  /**
+   * A list of values that have been pushed to the stream but not yet emitted to the user.
+   */
+  private queue: Array<ReadStreamResult<T, E>> = [];
+  /**
+   * Used by methods in the class to signal to the iterator that it
+   * should check for the next value.
+   */
+  private nextPromise: Promise<void> | null = null;
+  /**
+   * Resolves nextPromise
+   */
+  private resolveNextPromise: null | (() => void) = null;
+
+  public [Symbol.asyncIterator]() {
+    if (this.locked) {
+      throw new TypeError('Readable is already locked');
+    }
+
+    // first iteration with break signals an error, the following one signals end of iteration.
+    let didSignalDrain = false;
+    this.locked = true;
+
+    return {
+      next: async () => {
+        if (this.broken && didSignalDrain) {
+          return {
+            done: true,
+            value: undefined,
+          } as const;
+        }
+
+        // Wait until we have something in the queue
+        while (this.queue.length === 0) {
+          if (this.closed && !this.didBreakDisposeValues) {
+            return {
+              done: true,
+              value: undefined,
+            } as const;
+          }
+
+          if (this.broken) {
+            didSignalDrain = true;
+
+            return {
+              done: false,
+              value: Err(ReadableBrokenError),
+            } as const;
+          }
+
+          if (!this.nextPromise) {
+            this.nextPromise = new Promise<void>((resolve) => {
+              this.resolveNextPromise = resolve;
+            });
+          }
+
+          await this.nextPromise;
+          this.nextPromise = null;
+          this.resolveNextPromise = null;
+        }
+
+        // Unfortunately we have to use non-null assertion here, because T can be undefined
+        // we already check for array length above anyway
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const value = this.queue.shift()!;
+
+        return { done: false, value } as const;
+      },
+      return: () => {
+        this.break();
+        return { done: true, value: undefined } as const;
+      },
+    };
+  }
+
+  public async collect(): Promise<Array<ReadStreamResult<T, E>>> {
+    const array: Array<ReadStreamResult<T, E>> = [];
+    for await (const value of this) {
+      array.push(value);
+    }
+
+    return array;
+  }
+
+  public break(): undefined {
+    if (this.broken) {
+      return;
+    }
+
+    this.locked = true;
+    this.broken = true;
+    this.didBreakDisposeValues = this.queue.length > 0;
+    this.queue.length = 0;
+
+    // if we have any iterators waiting for the next value,
+    this.resolveNextPromise?.();
+  }
+
+  public isReadable(): boolean {
```

Comment:
```suggestion
  get readable(): boolean {
```

nit: in my mind readable almost always feels like a prop

Reply:
I'm not a fan of getters at all, I think they're terrible and should only be used when you have to use them (i.e. conforming to an interface that you have no control over). They're implicit function calls that obscure intention and inhibit the reader from knowing what is actually happening or what's possible.